File: frontend/src/physics/modes/triangle.ts
import type { PhysicsEngine } from "../PhysicsEngine";
import { ParticleSystem } from "../ParticleSystem";

export const triangleMode = {
  name: "triangle" as const,
  dim: 2 as const,

  defaultParams(): Record<string, number> {
    return {
      side_length: 100,
      spring_k: 0.3,
      angle_stiffness: 0.3,
      target_angle: Math.PI / 3,
      mass: 1,
      drag: 0.5,
      gravity_x: 0,
      gravity_y: 0,
      canvas_width: 800,
      canvas_height: 600,
    };
  },

  setup(engine: PhysicsEngine, params: Record<string, number>): void {
    const p = { ...this.defaultParams(), ...params };
    engine.reset();
    engine.dim = 2;
    engine.gravity = new Float32Array([p.gravity_x, p.gravity_y]);
    engine.dragCoefficient = p.drag;

    const cx = p.canvas_width / 2;
    const cy = p.canvas_height / 2;
    const side = p.side_length;

    const h = side * Math.sqrt(3) / 2;
    const p0 = [cx, cy - h * 2 / 3];
    const p1 = [cx + side / 2, cy + h / 3];
    const p2 = [cx - side / 2, cy + h / 3];

    engine.particles = ParticleSystem.create([p0, p1, p2]);

    const k = p.spring_k;
    engine.springs.add(0, 1, side, k);
    engine.springs.add(0, 2, side, k);
    engine.springs.add(1, 2, side, k);

    engine.angleConstraints.add(1, 0, 2, p.target_angle, p.angle_stiffness);

    engine.bounds = {
      minPos: new Float32Array([0, 0]),
      maxPos: new Float32Array([p.canvas_width, p.canvas_height]),
    };
    engine.boundsMode = "elastic";

    // Initial rotational velocity (counter-clockwise spin + drift)
    const dt = engine.dt;
    const ps = engine.particles;
    const spin = 80; // angular velocity → tangential speed
    // Centroid
    const centX = (p0[0] + p1[0] + p2[0]) / 3;
    const centY = (p0[1] + p1[1] + p2[1]) / 3;
    const drift = 40;
    const verts = [p0, p1, p2];
    for (let i = 0; i < 3; i++) {
      const dx = verts[i][0] - centX;
      const dy = verts[i][1] - centY;
      // Tangential velocity: perpendicular to radius, scaled by spin
      const vx = -dy / side * spin + drift;
      const vy = dx / side * spin;
      ps.prevPositions[i * 2] = ps.positions[i * 2] - vx * dt;
      ps.prevPositions[i * 2 + 1] = ps.positions[i * 2 + 1] - vy * dt;
    }
  },
};

File: backend/modes/triangle.py
from __future__ import annotations
import math
import torch
from .base import SimulationMode
from backend.physics.engine import PhysicsEngine
from backend.physics.particle import ParticleSystem
from backend.physics.collisions import Bounds


class TriangleMode(SimulationMode):
    """Three particles forming an equilateral triangle with position-based springs
    and an angle constraint. Verlet integration."""

    name = "triangle"
    dim = 2

    def default_params(self) -> dict:
        return {
            "side_length": 100.0,
            "spring_k": 0.3,
            "angle_stiffness": 0.3,
            "target_angle": math.pi / 3,
            "mass": 1.0,
            "drag": 0.5,
            "gravity_x": 0.0,
            "gravity_y": 0.0,
            "canvas_width": 800.0,
            "canvas_height": 600.0,
        }

    def setup(self, engine: PhysicsEngine, params: dict) -> None:
        p = {**self.default_params(), **params}
        engine.reset()
        engine.dim = self.dim
        engine.gravity = torch.tensor([p["gravity_x"], p["gravity_y"]])
        engine.drag_coefficient = p["drag"]

        cx = p["canvas_width"] / 2
        cy = p["canvas_height"] / 2
        side = p["side_length"]

        h = side * math.sqrt(3) / 2
        p0 = [cx, cy - h * 2 / 3]
        p1 = [cx + side / 2, cy + h / 3]
        p2 = [cx - side / 2, cy + h / 3]

        engine.particles = ParticleSystem.create(positions=[p0, p1, p2])

        k = p["spring_k"]
        for a, b in [(0, 1), (0, 2), (1, 2)]:
            engine.springs.add(a, b, side, k)

        engine.angle_constraints.add(1, 0, 2, p["target_angle"], p["angle_stiffness"])

        engine.bounds = Bounds(
            min_pos=torch.tensor([0.0, 0.0]),
            max_pos=torch.tensor([p["canvas_width"], p["canvas_height"]]),
        )
        engine.bounds_mode = "elastic"

        # Initial rotational velocity (counter-clockwise spin + drift)
        dt = engine.dt
        ps = engine.particles
        centx = (p0[0] + p1[0] + p2[0]) / 3
        centy = (p0[1] + p1[1] + p2[1]) / 3
        spin = 80.0
        drift = 40.0
        verts = [p0, p1, p2]
        for i in range(3):
            dx = verts[i][0] - centx
            dy = verts[i][1] - centy
            vx = -dy / side * spin + drift
            vy = dx / side * spin
            ps.prev_positions[i, 0] = ps.positions[i, 0] - vx * dt
            ps.prev_positions[i, 1] = ps.positions[i, 1] - vy * dt
